### PR TITLE
[Mailer] Fix the possibility to set a From header from MessageListener

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #31733
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Component/Mailer/SmtpEnvelope.php [hunk 1]
```diff
@@ -40,14 +37,7 @@ public function __construct(Address $sender, array $recipients)
 
     public static function create(RawMessage $message): self
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/SmtpEnvelope.php]: Not sure if we should keep this method.

**Comment by weaverryan** [comment]
@fabpot I'm getting a new error:

> An email must have a "From" header

I think it relates to the difference between "Sender" and "From" that I was mentioning (https://github.com/symfony/symfony/issues/31733#issuecomment-497456467) - though I'm far from an expert on this stuff. Should the listener set both sender *and* from?

**Comment by fabpot** [comment]
"From" is what you write on the letter, "Sender" is what you write on the envelope. So, if you don't provide an envelope, we create one for you based on what you wrote on the letter. So, you must always provide a "From" (or register a MessageListener to do it for you -- will be part of the semantic configuration when available of course).

File: src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @experimental in 4.3
+ *
+ * @internal
+ */
+final class DelayedSmtpEnvelope extends SmtpEnvelope
+{
+    private $senderSet;
+    private $recipientsSet;
```
Comment (derrabus) [inline src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: Since both fields are supposed to be booleans, let's initialize them with `false`.

Comment (fabpot) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: Done

File: src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ */
+final class DelayedSmtpEnvelope extends SmtpEnvelope
+{
+    private $senderSet = false;
+    private $recipientsSet = false;
+    private $message;
+
+    public function __construct(RawMessage $message)
+    {
+        if (!$message instanceof Message) {
```
Comment (xabbuh) [inline src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: Wouldn't it be better to change the type-hint instead? If we want to support any other message in the future we can still change the type then.

Comment (fabpot) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: I don't think so. I've done it that way to get a "better" error message.

Comment (xabbuh) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: But it means that the user will only get an error at runtime if they didn't discover before that `Message` instances are required here. With a proper type hint they can be warned before.

Comment (fabpot) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: How would they be warned before? I suppose you're thinking about static analysis. Not sure what to do.

Comment (fabpot) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: In any case, that's independent from this PR as I'm doing the same at several other places (and this code is copy/pasted from SmtpEnvelope).

Comment (fabpot) [reply src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: @xabbuh see #31837

**Pull request merged**